src/v/cloud_storage_clients/client_pool.cc

```diff
@@ -0,0 +1,135 @@
+    if (!_as.abort_requested()) {
+        _as.request_abort();
+    }
+    _cvar.broken();
+    // Wait until all leased objects are returned
+    co_await _gate.close();
+}
+
+void client_pool::shutdown_connections() {
+    _as.request_abort();
+    _cvar.broken();
+    for (auto& it : _leased) {
+        it.client->shutdown();
+    }
+    for (auto& it : _pool) {
+        it->shutdown();
+    }
+}
+
+/// \brief Acquire http client from the pool.
+///
+/// \note it's guaranteed that the client can only be acquired once
+///       before it gets released (release happens implicitly, when
+///       the lifetime of the pointer ends).
+/// \return client pointer (via future that can wait if all clients
+///         are in use)
+ss::future<client_pool::client_lease> client_pool::acquire() {
+    gate_guard guard(_gate);
+    try {
+        // If credentials have not yet been acquired, wait for them. It is
+        // possible that credentials are not initialized right after remote
+        // starts, and we have not had a response from the credentials API yet,
+        // but we have scheduled an upload. This wait ensures that when we call
+        // the storage API we have a set of valid credentials.
+        if (unlikely(!_apply_credentials)) {
+            co_await wait_for_credentials();
+        }
+
+        while (_pool.empty() && !_gate.is_closed() && !_as.abort_requested()) {
+            if (_policy == client_pool_overdraft_policy::wait_if_empty) {
+                co_await _cvar.wait();
+            } else {
+                _pool.emplace_back(make_client());
+            }
+        }
+    } catch (const ss::broken_condition_variable&) {
+    }
+    if (_gate.is_closed() || _as.abort_requested()) {
+        throw ss::gate_closed_exception();
+    }
+    vassert(!_pool.empty(), "'acquire' invariant is broken");
+    auto client = _pool.back();
+    _pool.pop_back();
+    client_lease lease(
+      client,
+      ss::make_deleter([pool = weak_from_this(), client, g = std::move(guard)] {
+          if (pool) {
+              pool->release(client);
+          }
+      }));
+    _leased.push_back(lease);
+    co_return lease;
+}
+
+size_t client_pool::size() const noexcept { return _pool.size(); }
+
+size_t client_pool::max_size() const noexcept { return _max_size; }
+
+void client_pool::populate_client_pool() {
+    for (size_t i = 0; i < _max_size; i++) {
+        _pool.emplace_back(make_client());
+    }
+}
+
+client_pool::http_client_ptr client_pool::make_client() const {
+    return std::visit(
+      [this](const auto& cfg) {
+          using cfg_type = std::decay_t<decltype(cfg)>;
+          static_assert(std::is_same_v<configuration, cfg_type>);
+          if constexpr (std::is_same_v<configuration, cfg_type>) {
+              return ss::make_shared<s3_client>(cfg, _as, _apply_credentials);
+          } else {
+              static_assert(always_false_v<cfg_type>, "Unknown client type");
+          }
+      },
```

Comment:
instead of type checking inside the std::visit it would be easier to read if asserting on the type held using accessors like std::holds_alternative<>.